### ISSUE TITLE
Realm force try error

### ISSUE THREAD

**Issue by goktugyil** [realm/SwiftLint#389]
https://realm.io/docs/swift/latest/

```swift
// Persist your data easily
let realm = try! Realm()
try! realm.write {
  realm.add(mydog)
}
```

I get the following error:
error: Force Try Violation: Force tries should be avoided. (force_try)

Its kind of ironic that Realm's first example in docs does not work with swiftlint. 

Not sure if I needed to open this issue here or somewhere else

**Comment by jpsim** [comment]
Not all uses of `try!` are bad, it's a rule in SwiftLint because it's sometimes a code smell.

In Realm's case, the initialization can never fail after the first initialization on a given thread, as explained in the [Error Handling](https://realm.io/docs/swift/latest/#error-handling) section of the docs:

> Like any disk IO operation, creating a Realm instance could sometimes fail if resources are constrained. In practice, this can only happen the first time a Realm instance is created on a given thread. Subsequent accesses to a Realm from the same thread will reuse a cached instance and will always succeed.

Which is why the use of `try!` in Realm's docs is appropriate.

**Comment by goktugyil** [comment]
I wasn't suggesting that it was bad or anything. Was just pointing out that I had to change .yml settings to compile Realm's example code. Which wasn't a good experience.

Maybe inside the rules, you could implement a default ignore for Realm related `try!` calls.

**Comment by Vramin** [comment]
I agree. This should either be a warning or the Realm examples should be rewritten with compliant syntax. I just started with SwiftLint and the first errors I get are based on this and on my line lengths. Neither should break my build.

**Comment by haydenholligan** [comment]
Does anybody else find it ironic that SwiftLint is maintained by Realm, and Realm's own code violates the SwiftLint guidelines?

**Comment by freak4pc** [comment]
You should probably exclude any external libraries as part of your linting process.

**Comment by haydenholligan** [comment]
Not the library itself, but the way Realm is designed to be written. When performing read/write transactions, you are supposed to use force tries (as noted above) to access the realm, which violates SwiftLint's default check

**Comment by freak4pc** [comment]
I just realised your issue with this :) a bit annoying I agree. 

To be honest for specifically Realm I have a wrapper similar to 

```swift
class MyRealms {
    /// swiftlint:disable:next force_try
    public static var main: Realm = try! Realm(configuration: ...)
}
```

That way I can just use `MyRealms.realm.write` or whichever and get rid of all of the warnings.

About writes etc, I would usually wrap those in a try/catch, just in case. so never had an issues with it.

This is just one thought of a way to go around this without disabling `force_try` on your entire code base, since it is a good lint rule in general :) 

By the way this seems more like a complain for realm/Realm then realm/SwiftLint :)

**Comment by jpsim** [comment]
> Does anybody else find it ironic that SwiftLint is maintained by Realm, and Realm's own code violates the SwiftLint guidelines?

I thought this went without saying, but clearly there's confusion here that I should address. SwiftLint provides stylistic hints, and there will always be a decent amount of "false positives" in a linter. Otherwise, if the rule applied 100% of the time, that syntax would be disallowed by the compiler.

In the case of Realm, there are language limitations in Swift that prevent the ideal use of context-sensitive error handling that Realm would ideally benefit from. Realm's error handling model is explained in the [official docs](https://realm.io/docs/swift/latest/#error-handling). There's a leaky abstraction there since the initializer can only fail in specific circumstances (e.g. first time it is invoked on a new thread) and Swift doesn't have a concept of errors that can only occurs in specific circumstances. Either something always or never `throws`. So Realm's use of `try!` is appropriate when you're familiar with these semantics. That being said, `try!` is still a code _smell_ (read, not an absolute error to be avoided at all costs), so SwiftLint reports it as such.

`try!` is a useful assertion, saying "hey Swift compiler, I know you think this can error, but I know better in this circumstance". When in doubt if an error can or cannot occur, sure handle errors. But otherwise, you owe it to yourself and your future codebase maintainers to document the expected flow of your program.

SwiftLint is _designed_ to be overruled by a human when the human knows better than the tool. Use `// swiftlint:disable` liberally. Don't make your code worse to please SwiftLint.

**Comment by haydenholligan** [comment]
I agree, I don't think there's anything wrong with SwiftLint in this case. I just disabled force tries.

**Comment by Vramin** [comment]
I just think it would save a lot of heartache and discussion if the default for this was a warning and not an error. It shouldn't break the build and send every first timer here wondering why it did. I'm sure more than a few of us only ran across SwiftLint because we're already using Realm so it's discordant for Realm example syntax to break under it. Like many here I've already fixed it and it won't trouble me again because I know it needs a workaround, but I'd like new adopters to have a better experience than I did.

**Comment by haydenholligan** [comment]
As discussed above, I think it's more of an issue for RealmSwift than SwiftLint. It's good style to avoid force tries, but RealmSwift uses them all the time.

**Comment by ogezue** [comment]
How should this be handled in reality? I don't want to have 30 warnings in my project and know that this is ok because easily I will oversee  warning 31. 
Isn't it possible to add an exclusion?

**Comment by freak4pc** [comment]
I suggest moving your realm creation to a singleton.l and add an exclusion only there. About writing, that could actually fail so I would usually put this in either a do/catch or with a try? If I don't care for the result.

**Comment by ogezue** [comment]
This is a good idea - thank you very much :)

**Comment by jpsim** [comment]
I strongly discourage moving Realm creation to a singleton, as that would bypass Realm's thread-specific creation logic. Sounds like a preferable solution would be to wrap Realm creation in a function that handles the common error handling you want to perform.

**Comment by freak4pc** [comment]
I had no issues with it thus far but that's an interesting though. 
Anyways my wrapping class also has a 

```swift
public static func makeRealm() -> Realm {
    return try! Realm(configuration: THRealms.config)
}
```

Which is similar to what you suggested.

**Comment by ogezue** [comment]
Anything bad with my RealmProvider class?

```
class RealmProvider {

    private var realm: Realm?
    private let currentSchemaVersion: UInt64 = 5

    func loadRealm() -> Realm {

        if let realm = self.realm {
            return realm
        }

        migrateRealm()

        if let _ = NSClassFromString("XCTest") {
            realm =  try! Realm(configuration: Realm.Configuration(fileURL: nil, inMemoryIdentifier: "test", syncConfiguration: nil, encryptionKey: nil, readOnly: false, schemaVersion: currentSchemaVersion, migrationBlock: nil, deleteRealmIfMigrationNeeded: true, objectTypes: nil))
        } else {
            realm = try! Realm()
        }
        return realm!
    }

    func migrateRealm() {

        Realm.Configuration.defaultConfiguration = Realm.Configuration(
            schemaVersion: currentSchemaVersion, migrationBlock: { _, oldSchemaVersion in

                if (oldSchemaVersion < self.currentSchemaVersion) {
                    self.purgeRealm()
                }
        })
    }

    private func purgeRealm() {
        logger.warning("deleting realm data...")
        do {
            if let realmUrl = Realm.Configuration.defaultConfiguration.fileURL {
                try FileManager.default.removeItem(at: realmUrl)
            }
        } catch {
            logger.error("failed deleting realm data")
        }
    }
```

**Comment by jpsim** [comment]
@goktugyil this probably isn't the best place to get analysis on your usage of Realm. Try asking a question on StackOverflow with the `realm` tag? http://stackoverflow.com/questions/ask?tags=realm

**Comment by goktugyil** [comment]
@jpsim  I think you meant to tag @ogezue

**Comment by jpsim** [comment]
oh yes, sorry for the ping.
